src/freenet/support/io/InetAddressIpv6FirstComparator.java

```diff
@@ -0,0 +1,92 @@
+package freenet.support.io;
+
+import static freenet.node.NodeStats.DEFAULT_MAX_PING_TIME;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.Comparator;
+
+import freenet.support.Fields;
+import freenet.support.LRUCache;
+
+/** Comparator for IP addresses that sorts IPv6 before IPv4 to enable
+ * selecting the first.
+ * @author toad */
+public class InetAddressIpv6FirstComparator implements Comparator<InetAddress> {
+
+	// need a cache for reachability to avoid doing NlogN issReachable checks in worst case.
+	public LRUCache<Integer, Boolean> reachabilityCache = new LRUCache<>(1000, 300000);
+
+	public static final InetAddressIpv6FirstComparator COMPARATOR =
+		new InetAddressIpv6FirstComparator();
+
+	@Override
+	public int compare(InetAddress arg0, InetAddress arg1) {
+		if(arg0 == arg1) return 0;
+		// prefer everything over broadcast
+		if (!arg0.isAnyLocalAddress() && arg1.isAnyLocalAddress()) {
+			return -1;
+		} else if (arg0.isAnyLocalAddress() && !arg1.isAnyLocalAddress()) {
+			return 1;
+		}
+		// prefer everything over loopback / localhost
+		if (!arg0.isLoopbackAddress() && arg1.isLoopbackAddress()) {
+			return -1;
+		} else if (arg0.isLoopbackAddress() && !arg1.isLoopbackAddress()) {
+			return 1;
+		}
+		// prefer LAN routable addresses over link-local addresses because most of these will not work
+		if (!arg0.isLinkLocalAddress() && arg1.isLinkLocalAddress()) {
+			return -1;
+		} else if (arg0.isLinkLocalAddress() && !arg1.isLinkLocalAddress()) {
+			return 1;
+		}
+		// prefer reachable over unreachable addresses. This is usually a ping. TODO: This actually pings all advertised ip addresses. Is that OK? Do we need a maximum number of accepted addresses to prevent abuse as DDoS? 
+		int a = arg0.hashCode();
+		int b = arg1.hashCode();
+		Boolean reachable0 = reachabilityCache.get(a);
+		Boolean reachable1 = reachabilityCache.get(b);
+		if (reachable0 == null) {
+			try {
+				reachable0 = arg0.isReachable((int) DEFAULT_MAX_PING_TIME);
```

Comment:
Is it reasonable to not allow configuring this ping time? I'm not sure about the usability and security implications here, but wanted to note that this is hardcoded.

Reply:
I think it’s reasonable. I hope there won’t be a need to configure it, but if that arises, we can add that.